File: android/RecyclerView-master/GithubUserSub3/app/src/main/java/com/dicoding/githubusersub3/layout/FavoriteActivity.java
package com.dicoding.githubusersub3.layout;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;

import com.dicoding.githubusersub3.R;
import com.dicoding.githubusersub3.adapter.FavoriteAdapter;
import com.dicoding.githubusersub3.data.User;
import com.dicoding.githubusersub3.database.UserHelper;

import java.util.ArrayList;

public class FavoriteActivity extends AppCompatActivity {

    private UserHelper userHelper;
    private ArrayList<User> userList = new ArrayList<>();
    private FavoriteAdapter favoriteAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_favorite);
        userHelper = new UserHelper(getApplicationContext());
        userHelper.open();
        userList = userHelper.getDataUser();
        setRecyclerView();
        if (getActionBar() != null) {
            getActionBar().hide();
        }
    }

    private void setRecyclerView() {
        RecyclerView recyclerView = findViewById(R.id.rv_fav);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setHasFixedSize(true);
        favoriteAdapter = new FavoriteAdapter(getApplication());
        recyclerView.setAdapter(favoriteAdapter);
    }

    @Override
    protected void onResume() {
        super.onResume();
        userList = userHelper.getDataUser();
        favoriteAdapter.setUserGithubList(userList);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        userHelper.close();
    }
}

File: android/RecyclerView-master/GithubUserFix/app/src/main/java/com/dicoding/githubuserfix/layout/FavoriteActivity.kt
package com.dicoding.githubuserfix.layout

import android.database.ContentObserver
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.Handler
import android.os.HandlerThread
import android.view.View
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import com.dicoding.githubuserfix.R
import com.dicoding.githubuserfix.adapter.FavoriteAdapter
import com.dicoding.githubuserfix.data.FavoriteData
import com.dicoding.githubuserfix.database.DbContract.FavColumns.Companion.CONTENT_URI
import com.dicoding.githubuserfix.helper.MappingHelper
import kotlinx.android.synthetic.main.activity_favorite.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.launch

class FavoriteActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_favorite)
        setActionBarTitle()
        rvFavorite.layoutManager = LinearLayoutManager(this)
        rvFavorite.setHasFixedSize(true)
        adapter = FavoriteAdapter(this)
        rvFavorite.adapter = adapter
        val handlerThread = HandlerThread("DataObserver")
        handlerThread.start()
        val handler = Handler(handlerThread.looper)
        val myObserver = object : ContentObserver(handler) {
            override fun onChange(self: Boolean) {
                loadNotesAsync()
            }
        }
        contentResolver.registerContentObserver(CONTENT_URI, true, myObserver)
        if (savedInstanceState == null) {
            loadNotesAsync()
        } else {
            val list = savedInstanceState.getParcelableArrayList<FavoriteData>(EXTRA_STATE)
            if (list != null) {
                adapter.listFavorite = list
            }
        }
    }

    private fun loadNotesAsync() {
        GlobalScope.launch(Dispatchers.Main) {
            progressBarFavorite.visibility = View.VISIBLE
            val defNotes = async(Dispatchers.IO) {
                val cursor = contentResolver?.query(CONTENT_URI, null, null, null, null)
                MappingHelper.mapCursorToArrayList(cursor)
            }
            val favoriteData = defNotes.await()
            progressBarFavorite.visibility = View.INVISIBLE
            if (favoriteData.size > 0) {
                adapter.listFavorite = favoriteData
            } else {
                adapter.listFavorite = ArrayList()
                showMessage()
            }
        }
    }

    private fun showMessage() {
        Toast.makeText(this, getString(R.string.empty_favorite), Toast.LENGTH_SHORT).show()
    }

    private fun setActionBarTitle() {
        if (supportActionBar != null) {
            supportActionBar?.title = "Favorite Users"
        }
    }

    override fun onResume() {
        super.onResume()
        loadNotesAsync()
    }

    private lateinit var adapter: FavoriteAdapter

    companion object {
        private const val EXTRA_STATE = "extra_state"
    }
}